File: src/main/java/org/example/flink/scala/WordCountStream.scala
package org.example.flink.scala


import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment


/**
 * WordCountStream对象实现了通过流处理计算文本中单词出现次数的示例。
 * 它使用Flink的流处理API来从指定的socket流中读取数据，进行单词计数，并打印结果。
 */
object WordCountStream {

  /**
   * 程序的入口点。
   * @param args 命令行参数，目前未使用。
   */
  def main(args: Array[String]): Unit = {
    // 获取流执行环境，这是Flink流处理程序的起点。
    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment

    // 设置并行度为1，确保程序顺序执行，便于调试。
    env.setParallelism(1)

    // 导入Flink的Scala API，用于操作DataStream。
    import org.apache.flink.streaming.api.scala._

    // 从指定的socket地址和端口读取文本流。
    // 这里的socketTextStream用于模拟数据输入，通常在实际场景中会连接到实际的数据源。
    val stream: DataStream[String] = env.socketTextStream("node3", 8888)

    // 对文本流进行单词计数的处理流程：
    // 1. 使用flatMap将每行文本拆分成单词。
    // 2. 使用map为每个单词生成一个元组，形式为(word, 1)，表示该单词出现一次。
    // 3. 使用keyBy将数据按照单词进行分组，为每个分组内的单词进行累加。
    // 4. 使用sum对每个分组内的单词出现次数进行累加。
    val result: DataStream[(String, Int)] = stream.flatMap(_.split("\\s+"))
      .map((_, 1))
      .keyBy(_._1)
      .sum(1)

    // 将结果打印到控制台。
    result.print()

    // 启动Flink作业执行。
    env.execute("wordcount")
  }
}

File: src/main/java/org/example/flink/java/WordCountStream.java
package org.example.flink.java;

import org.apache.flink.api.common.functions.FlatMapFunction;
import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.datastream.KeyedStream;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.util.Collector;

/**
 * WordCount流处理程序的实现类。
 * 该程序演示了如何使用Apache Flink进行简单的单词计数流处理。
 */
public class WordCountStream {
    /**
     * 程序的入口点。
     * @param args 命令行参数，预期使用节点地址和端口作为输入源。
     * @throws Exception 如果执行环境设置或操作失败。
     */
    public static void main(String[] args) throws Exception {
        // 创建流执行环境
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        // 设置并行度为1，确保程序顺序执行
        env.setParallelism(1);

        // 从指定的socket读取文本流
        DataStreamSource<String> dataSource = env.socketTextStream("node3",8888);

        // 对文本流进行分词，每个单词作为一个单独的元素
        SingleOutputStreamOperator<String> words = dataSource.flatMap(new FlatMapFunction<String, String>() {
            @Override
            public void flatMap(String s, Collector<String> collector) throws Exception {
                String [] words = s.split("\\s+");
                for (String word: words) {
                    collector.collect(word);
                }
            }
        });

        // 将每个单词与计数1关联，准备进行聚合操作
        SingleOutputStreamOperator<Tuple2<String, Integer>> tupSteram = words.map(new MapFunction<String, Tuple2<String, Integer>>() {
            @Override
            public Tuple2<String, Integer> map(String s) throws Exception {
                return Tuple2.of(s, 1);
            }
        });

        // 根据单词进行分组，为单词计数做准备
        KeyedStream<Tuple2<String, Integer>, String> keyedStream = tupSteram.keyBy(tup -> tup.f0);

        // 计算每个单词出现的总次数
        SingleOutputStreamOperator<Tuple2<String, Integer>> sum = keyedStream.sum(1);

        // 将结果打印到控制台
        sum.print();

        // 执行流处理作业
        env.execute("WordCountStream");
    }
}
